### PR TITLE
ARROW-5205: [Python] Improved error messages when user erroneously uses a non-local resource URI to open a file

### PR DESCRIPTION
**Pull request by nealrichardson**
Here's a solution for [the issue](https://issues.apache.org/jira/browse/ARROW-5205) that addresses not just the [original report](https://stackoverflow.com/questions/55704943/what-could-be-the-explanation-of-this-pyarrow-lib-arrowioerror/55707311#55707311) about `read_csv` but presumably any other way you could try to load a non-local file. 

I could think of a few other ways to solve this, and we can certainly debate what the better error messages should say, but figured it's easier to discuss in the context of a patch. I also wonder whether pyarrow should just try to do itself what the more helpful error message suggests (i.e. if you get a hdfs path string and it fails, try to connect with HadoopFileSystem and retry rather than tell the user to retry) but maybe that's too clever.

### CODE REVIEW

File: python/pyarrow/tests/test_csv.py [hunk 1]
```diff
@@ -59,6 +59,13 @@ def make_random_csv(num_cols=2, num_rows=10, linesep=u'\r\n'):
     return csv, expected
 
 
+def test_nonlocal_file_errors_nicely():
+    with pytest.raises(pa.lib.ArrowIOError, match="Try using HadoopFileSystem"):
+        read_csv("hdfs://does-not-exist.csv")
+    with pytest.raises(pa.lib.ArrowIOError, match="Try using S3FSWrapper"):
+        read_csv("s3://does-not-exist.csv")
```
Comment (wesm) [inline python/pyarrow/tests/test_csv.py]: Neither of these classes are really public APIs, `HadoopFileSystem` is obtained through `pa.hdfs.connect` and `S3FSWrapper` is an implementation detail when the user passes a s3fs file handle. 

A better solution might be to parse the URI and see if there is a "scheme" at all and say that `URI scheme $scheme cannot be accessed on the local filesystem`

Comment (wesm) [reply python/pyarrow/tests/test_csv.py]: @pitrou thoughts?

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: Yes, I think rejecting URIs upfront would be better than this.

Comment (nealrichardson) [reply python/pyarrow/tests/test_csv.py]: Particularly given https://github.com/apache/arrow/pull/4225, might the right place to reject up front be down in the C++ layer?

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: That's an interesting idea at least.

**Comment by pitrou** [comment]
Should we keep this PR open?

**Comment by nealrichardson** [comment]
Happy to close since it doesn't appear that this is the right direction, but while we're at it, can we agree on what the right solution is so I can update the Jira issue? Should this be handled in the C++ filesystem API more generally? Or should there be pyarrow specific handling? IMO if we're not going to offer a Python-specific way to solve/retry when this error is hit, there's no reason it shouldn't get pushed down to C++.

**Comment by pitrou** [comment]
Not sure yet. We will probably have an API on the C++ side to handle URIs and dispatch them to the proper filesystem implementation, I'd say.

**Pull request closed without merging**